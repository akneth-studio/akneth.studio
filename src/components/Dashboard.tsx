import { createClient } from '@supabase/supabase-js'
import Link from 'next/link';

const supabase = createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!
)

export default async function Dashboard() {
    const { data: messages, error } = await supabase
        .from('messages')
        .select('id, name, company, email, subject, message, created_at, replied, reply_text')
        .order('created_at', { ascending: false })
        .limit(5)

    if (error) {
        return <div className="alert alert-danger">Błąd ładowania wiadomości: {error.message}</div>
    }

    return (
        <section className='messg'>
            <h2 className="mb-4">Wiadomości z formularza kontaktowego</h2>
            {(!messages || messages.length === 0) ? (
                <div className="alert alert-info">Brak wiadomości.</div>
            ) : (
                <div className="table-responsive">
                    <table className="table table-striped align-middle">
                        <thead>
                            <tr>
                                <th>Data</th>
                                <th>Imię i nazwisko</th>
                                <th>Firma</th>
                                <th>E-mail</th>
                                <th>Temat</th>
                                <th>Treść</th>
                                <th>Odpowiedziano?</th>
                                <th>Treść odpowiedzi</th>
                            </tr>
                        </thead>
                        <tbody>
                            {messages.map(msg => (
                                <tr key={msg.id}>
                                    <td style={{ whiteSpace: 'nowrap' }}>
                                        {msg.created_at
                                            ? new Date(msg.created_at).toLocaleString('pl-PL', {
                                                day: '2-digit',
                                                month: '2-digit',
                                                year: 'numeric',
                                                hour: '2-digit',
                                                minute: '2-digit'
                                            })
                                            : '—'}
                                    </td>
                                    <td>{msg.name || <span className="text-muted">—</span>}</td>
                                    <td>{msg.company || <span className="text-muted">—</span>}</td>
                                    <td>
                                        {msg.email ? (
                                            <a href={`mailto:${msg.email}`}>{msg.email}</a>
                                        ) : (
                                            <span className="text-muted">—</span>
                                        )}
                                    </td>
                                    <td>{msg.subject || <span className="text-muted">—</span>}</td>
                                    <td style={{ maxWidth: 320, whiteSpace: 'pre-line', wordBreak: 'break-word' }}>
                                        {msg.message || <span className="text-muted">—</span>}
                                    </td>
                                    <td>
                                        {msg.replied
                                            ? <span className="badge bg-success">Tak</span>
                                            : <span className="badge bg-secondary">Nie</span>
                                        }
                                    </td>
                                    <td style={{ maxWidth: 320, whiteSpace: 'pre-line', wordBreak: 'break-word' }}>
                                        {msg.reply_text || <span className="text-muted">—</span>}
                                    </td>
                                </tr>
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
            <div className="text-end mt-3">
                <Link href="/admin/messages" className="btn btn-outline-primary">
                    Więcej &rarr;
                </Link>
            </div>
        </section>
    )
}
