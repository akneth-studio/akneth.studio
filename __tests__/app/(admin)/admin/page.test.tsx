/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import Dashboard from '@/app/(admin)/admin/page';

// Mock environment variables
process.env.NEXT_PUBLIC_SITE_URL = require('../../../../jest.setup').NEXT_PUBLIC_SITE_URL || 'https://akneth.pl';

// Mock the CTAButton component
jest.mock('@/components/CTAButton', () => {
  return {
    __esModule: true,
    default: jest.fn(({ text, to }) => {
      return React.createElement('button', { 'data-testid': 'cta-button', 'data-text': text, 'data-to': to }, text);
    })
  };
});

// Mock the Summary component
jest.mock('@/components/admin/Summary', () => {
  return {
    __esModule: true,
    default: jest.fn(() => {
      return React.createElement('div', { 'data-testid': 'summary-component' }, 'Summary Component');
    })
  };
});

// Mock Supabase client
jest.mock('@/utils/supabase/client', () => {
  return {
    supabase: {
      auth: {
        getUser: jest.fn(),
      },
    },
  };
});

describe('Admin Dashboard Page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    // Reset the mock implementation
    const { supabase } = require('@/utils/supabase/client');
    supabase.auth.getUser.mockReset();
  });

  it('should render login prompt when user is not authenticated', async () => {
    // Mock unauthenticated user
    const { supabase } = require('@/utils/supabase/client');
    supabase.auth.getUser.mockResolvedValue({ data: { user: null } });

    render(<Dashboard />);
    
    // Wait for useEffect to complete
    await waitFor(() => {
      expect(screen.getByText((content, element) => {
        const hasText = (node: Element | null) => node !== null && node.textContent === 'Witaj, Użytkowniku!';
        const elementHasText = hasText(element);
        const childrenDontHaveText = Array.from(element?.children || []).every(child => !hasText(child));
        return elementHasText && childrenDontHaveText;
      })).toBeInTheDocument();
    });
    
    expect(screen.getByText('Dostęp tylko po zalogowaniu.')).toBeInTheDocument();
    
    // Check if login button is rendered
    const loginButtons = screen.getAllByTestId('cta-button');
    expect(loginButtons[0]).toBeInTheDocument();
    expect(loginButtons[0]).toHaveAttribute('data-text', 'Zaloguj się');
    expect(loginButtons[0]).toHaveAttribute('data-to', '/admin/login');
  });

  it('should render admin panel when user is authenticated', async () => {
    // Mock authenticated user
    const mockUser = {
      id: 'user123',
      user_metadata: {
        display_name: 'Admin User'
      }
    };
    
    const { supabase } = require('@/utils/supabase/client');
    supabase.auth.getUser.mockResolvedValue({ data: { user: mockUser } });

    render(<Dashboard />);
    
    // Wait for useEffect to complete
    await waitFor(() => {
      expect(screen.getByText('PANEL ADMINA')).toBeInTheDocument();
    });
    
    expect(screen.getByText((content, element) => {
      const hasText = (node: Element | null) => node !== null && node.textContent === 'Witaj, Admin User!';
      const elementHasText = hasText(element);
      const childrenDontHaveText = Array.from(element?.children || []).every(child => !hasText(child));
      return elementHasText && childrenDontHaveText;
    })).toBeInTheDocument();
    expect(screen.getByText('Jesteś zalogowany jako admin.')).toBeInTheDocument();
    expect(screen.getByTestId('summary-component')).toBeInTheDocument();
  });

  it('should render with default name when display_name is not available', async () => {
    // Mock authenticated user without display_name
    const mockUser = {
      id: 'user123',
      user_metadata: {}
    };
    
    const { supabase } = require('@/utils/supabase/client');
    supabase.auth.getUser.mockResolvedValue({ data: { user: mockUser } });

    render(<Dashboard />);
    
    // Wait for useEffect to complete
    await waitFor(() => {
      expect(screen.getByText((content, element) => {
        const hasText = (node: Element | null) => node !== null && node.textContent === 'Witaj, Użytkowniku!';
        const elementHasText = hasText(element);
        const childrenDontHaveText = Array.from(element?.children || []).every(child => !hasText(child));
        return elementHasText && childrenDontHaveText;
      })).toBeInTheDocument();
    });
  });

  it('should match snapshot when user is not authenticated', async () => {
    const { supabase } = require('@/utils/supabase/client');
    supabase.auth.getUser.mockResolvedValue({ data: { user: null } });

    const { container } = render(<Dashboard />);
    
    // Wait for useEffect to complete
    await waitFor(() => {
      expect(screen.getByText((content, element) => {
        const hasText = (node: Element | null) => node !== null && node.textContent === 'Witaj, Użytkowniku!';
        const elementHasText = hasText(element);
        const childrenDontHaveText = Array.from(element?.children || []).every(child => !hasText(child));
        return elementHasText && childrenDontHaveText;
      })).toBeInTheDocument();
    });
    
    expect(container).toMatchSnapshot();
  });
});
