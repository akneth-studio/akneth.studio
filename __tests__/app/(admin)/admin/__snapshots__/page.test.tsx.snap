// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Admin Dashboard Page should match snapshot when user is not authenticated 1`] = `
<div>
  <div
    class="text-center"
  >
    Witaj, 
    <span
      class="fw-bold"
    >
      Użytkowniku
    </span>
    !
  </div>
  <div
    class="align-items-center"
  >
    <h2
      class="text-center"
    >
      Dostęp tylko po zalogowaniu.
    </h2>
    <div
      class="text-center"
    >
      <button
        data-testid="cta-button"
        data-text="Zaloguj się"
        data-to="/admin/login"
      >
        Zaloguj się
      </button>
      <button
        data-testid="cta-button"
        data-text="Strona główna"
        data-to="/"
      >
        Strona główna
      </button>
    </div>
  </div>
</div>
`;
